collect_app/src/main/java/org/odk/collect/android/http/OkHttpConnection.java

```diff
@@ -0,0 +1,352 @@
+
+            MimeTypeMap mimeTypeMap = MimeTypeMap.getSingleton();
+
+            long byteCount = 0L;
+
+            RequestBody requestBody = RequestBody.create(MediaType.parse(HTTP_CONTENT_TYPE_TEXT_XML), submissionFile);
+
+            MultipartBody.Builder multipartBuilder = new MultipartBody.Builder()
+                    .setType(MultipartBody.FORM)
+                    .addPart(MultipartBody.Part.createFormData("xml_submission_file", submissionFile.getName(), requestBody));
+
+            Timber.i("added xml_submission_file: %s", submissionFile.getName());
+            byteCount += submissionFile.length();
+
+            for (; fileIndex < fileList.size(); fileIndex++) {
+                File file = fileList.get(fileIndex);
+
+                String mime = mimeTypeMap.getMimeTypeFromExtension(FileUtils.getFileExtension(file.getName()));
+
+                RequestBody fileRequestBody = RequestBody.create(MediaType.parse(mime), file);
+
+                multipartBuilder.addPart(MultipartBody.Part.create(fileRequestBody));
+
+                byteCount += file.length();
+                Timber.i("added file of type '%s' %s", mime, file.getName());
+
+                // we've added at least one attachment to the request...
+                if (fileIndex + 1 < fileList.size()) {
+                    if ((fileIndex - lastFileIndex + 1 > 100) || (byteCount + fileList.get(fileIndex + 1).length()
+                            > 10000000L)) {
+                        // the next file would exceed the 10MB threshold...
+                        Timber.i("Extremely long post is being split into multiple posts");
+                        multipartBuilder.addPart(MultipartBody.Part.createFormData("*isIncomplete*", "yes"));
+                        ++fileIndex; // advance over the last attachment added...
+                        break;
+                    }
+                }
+            }
+
+            MultipartBody multipartBody = multipartBuilder.build();
+
+            try {
+                OkHttpClient client = createOkHttpClient(credentials);
+                Request request = postRequest(uri, multipartBody);
+                Response response = client.newCall(request).execute();
+
+                messageParser = new ResponseMessageParser(
+                        response.toString(),
+                        response.code(),
+                        response.message());
+
+                discardEntityBytes(response);
+
+                if (response.code() != HttpURLConnection.HTTP_CREATED && response.code() != HttpURLConnection.HTTP_ACCEPTED) {
+                    return messageParser;
+                }
+            } catch (IOException e) {
+                Timber.e(e);
+                String msg = e.getMessage();
+                if (msg == null) {
+                    msg = e.toString();
+                }
+
+                throw new IOException(msg);
+            }
+        }
+
+        return messageParser;
```

Comment:
Wondering whether `ResponseMessageParser` should really be renamed to `HttpPostResult` by analogy to the other two methods above. And I also believe any HTTP response could provide a custom `message` though of course it's much more useful in response to a post: https://docs.opendatakit.org/openrosa-http/#http-responses. That could be filed as a separate issue.

Reply:
See comment above, I refactored ResponseMessageParser and created a HttpPostResult.